src/renderers/opengl/app.rs

```diff
@@ -0,0 +1,169 @@
+                .unwrap_or_else(|_| {
+                    display
+                        .create_context(&config, &fallback_context_attributes)
+                        .expect("failed to create context")
+                })
+        };
+
+        let raw_window_handle = window.raw_window_handle();
+
+        let context_attributes = glutin::context::ContextAttributesBuilder::new()
+            .with_profile(glutin::context::GlProfile::Compatibility)
+            .with_context_api(glutin::context::ContextApi::Gles(Some(
+                glutin::context::Version::new(2, 0),
+            )))
+            .build(Some(raw_window_handle));
+
+        let dimensions = window.inner_size();
+
+        let (surface, gl_ctx) = {
+            let attrs = SurfaceAttributesBuilder::<glutin::surface::WindowSurface>::new().build(
+                raw_window_handle,
+                NonZeroU32::new(dimensions.width).unwrap(),
+                NonZeroU32::new(dimensions.height).unwrap(),
+            );
+
+            let surface = unsafe { display.create_window_surface(&config, &attrs)? };
+            let context = unsafe { display.create_context(&config, &context_attributes)? }
+                .make_current(&surface)?;
+            (surface, context)
+        };
+
+        // Load the OpenGL function pointers
+        let gl = unsafe {
+            glow::Context::from_loader_function(|symbol| {
+                display.get_proc_address(&CString::new(symbol).unwrap()) as *const _
+            })
+        };
+
+        unsafe {
+            gl.debug_message_callback(|_src, ty, _id, sevr, msg| {
+                let ty = match ty {
+                    glow::DEBUG_TYPE_ERROR => "Error: ",
+                    glow::DEBUG_TYPE_DEPRECATED_BEHAVIOR => "Deprecated Behavior: ",
+                    glow::DEBUG_TYPE_MARKER => "Marker: ",
+                    glow::DEBUG_TYPE_OTHER => "",
+                    glow::DEBUG_TYPE_POP_GROUP => "Pop Group: ",
+                    glow::DEBUG_TYPE_PORTABILITY => "Portability: ",
+                    glow::DEBUG_TYPE_PUSH_GROUP => "Push Group: ",
+                    glow::DEBUG_TYPE_UNDEFINED_BEHAVIOR => "Undefined Behavior: ",
+                    glow::DEBUG_TYPE_PERFORMANCE => "Performance: ",
+                    ty => unreachable!("unknown debug type {ty}"),
+                };
+                match sevr {
+                    glow::DEBUG_SEVERITY_NOTIFICATION => debug!(target: "opengl", "{ty}{msg}"),
+                    glow::DEBUG_SEVERITY_LOW => info!(target: "opengl", "{ty}{msg}"),
+                    glow::DEBUG_SEVERITY_MEDIUM => warn!(target: "opengl", "{ty}{msg}"),
+                    glow::DEBUG_SEVERITY_HIGH => error!(target: "opengl", "{ty}{msg}"),
+                    sevr => unreachable!("unknown debug severity {sevr}"),
+                };
+            });
+
+            gl.enable(glow::DEBUG_OUTPUT);
+        }
+
+        unsafe { gl.viewport(0, 0, 2048, 2048) };
```

Comment:
Just a note to myself that I have to do something about that viewport. Inochi2D has a `Camera` class to handle puppet position and zoom, but I have none of that currently, I'll have to see what I can do.

Reply:
Yup! And also, I have been wondering. We need a way to render multiple puppets within one camera. 

Especially considering my use case would be for an anime RPG game with puppet rendering as sprites in the dialogs, akin to the one in Hatsune Miku: Colorful Stage.

Reply:
If we get a Camera struct, then I think we'll get that for free.